xtask/src/semver_check.rs

```diff
@@ -0,0 +1,278 @@
+
+            let current_path = build_doc_json(package, chip, &package_path)?;
+
+            let suffix = if package.chip_features_matter() {
+                format!("_{}", chip.to_string())
+            } else {
+                "".to_string()
+            };
+
+            let baseline_path = package_path.join(format!("api_baseline{}.json", suffix));
+
+            remove_unstable_items(&current_path)?;
+
+            let mut check = cargo_semver_checks::Check::new(
+                cargo_semver_checks::Rustdoc::from_path(current_path),
+            );
+            check.set_baseline(cargo_semver_checks::Rustdoc::from_path(baseline_path));
+            let mut cfg = cargo_semver_checks::GlobalConfig::new();
+            // seems there is no way of getting the details via the API - so lets log them
+            cfg.set_log_level(Some(log::Level::Info));
+            let check_result = check.check_release(&mut cfg)?;
+
+            log::info!("Result {:?}", check_result);
+
+            for (_, report) in check_result.crate_reports() {
+                if let Some(required_bump) = report.required_bump() {
+                    if required_bump == cargo_semver_checks::ReleaseType::Major {
+                        return Err(anyhow::anyhow!(
+                            "Semver check failed: {:?}",
+                            report.required_bump()
+                        ));
+                    }
+                }
+            }
+
+            if !package.chip_features_matter() {
+                break;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn build_doc_json(
+    package: Package,
+    chip: &Chip,
+    package_path: &PathBuf,
+) -> Result<PathBuf, anyhow::Error> {
+    let current_path = PathBuf::from(package_path)
+        .join("target")
+        .join(chip.target())
+        .join("doc")
+        .join(format!("{}.json", package.to_string().replace("-", "_")));
+    std::fs::remove_file(&current_path).ok();
+    let toolchain = "esp";
+    let features = if package.chip_features_matter() {
+        vec![chip.to_string(), "unstable".to_string()]
+    } else {
+        vec!["unstable".to_string()]
+    };
+    let mut cargo_builder = CargoArgsBuilder::default()
+        .toolchain(toolchain)
+        .subcommand("rustdoc")
+        .features(&features)
+        .target(chip.target())
+        .arg("-Zunstable-options")
+        .arg("--lib")
+        .arg("--output-format=json");
+    cargo_builder = cargo_builder.arg("-Zbuild-std=alloc,core");
+    let cargo_args = cargo_builder.build();
+    log::debug!("{cargo_args:#?}");
+    crate::cargo::run(&cargo_args, package_path)?;
+    Ok(current_path)
+}
+
+fn remove_unstable_items(path: &Path) -> Result<(), anyhow::Error> {
+    // this leaves orphaned items! cargo-semver-checks seems to be fine with that however
+    // the json fmt is unstable - we might fail when using the "wrong" version of `rustdoc_types` here
```

Comment:
This concerns me a bit, it would be really nice to leave all the rustdoc specific output parsing to the semverchecks library, but I guess this is our only option for now?

Reply:
Ideally `instability::unstable` wouldn't unconditionally emit `#[cfg(any(doc, feature = "unstable"))]` (the doc - part) - then we could generate a suitable doc-json with just `rustdoc` and remove this whole function

I hope we can do that in future